dali/operators/image/crop/bbox_crop.cc

```diff
@@ -14,143 +14,635 @@
+        }
+      } else {  // relative dimensions
+        std::uniform_real_distribution<float> extent_dist(scale_range_.min, scale_range_.max);
+        for (int d = 0; d < ndim; d++) {
+          shape[d] = extent_dist(rng);
+          out_bounds[ndim + d] = rel_bounds[ndim + d] = shape[d];
+        }
+
+        if (!ValidAspectRatio(make_cspan(shape))) {
+          continue;
+        }
+
+        for (int d = 0; d < ndim; d++) {
+          std::uniform_real_distribution<float> anchor_dist(0.0f, 1.0f - shape[d]);
+          anchor[d] = anchor_dist(rng);
+          out_bounds[d] = rel_bounds[d] = anchor[d];
+        }
+      }
+
+      auto layout = DefaultAnchorAndShapeLayout<ndim>();  // start and shape
+      Box<ndim, float> rel_crop;
+      ReadBox(rel_crop, make_cspan(rel_bounds), layout);
+
+      Box<ndim, float> out_crop;
+      ReadBox(out_crop, make_cspan(out_bounds), layout, {});
+
+      if (!ValidOverlap(rel_crop, bounding_boxes, option.min_iou)) {
+        continue;
+      }
 
-  BoundingBoxes bounding_boxes;
-  bounding_boxes.reserve(static_cast<size_t>(boxes_tensor.dim(0)));
+      auto filtered_bboxes = bounding_boxes;
+      auto filtered_labels = labels;
+      FilterByCentroid(rel_crop, filtered_bboxes, filtered_labels);
+      if (filtered_bboxes.empty()) {
+        continue;
+      }
 
-  for (int i = 0; i < boxes_tensor.dim(0); ++i) {
-    const auto *box_data =
-        boxes_tensor.data<float>() + (i * BoundingBox::kSize);
+      for (auto &box : filtered_bboxes) {
+        box = RemapBox(box, rel_crop);
+      }
 
-    auto box = ltrb_ ? BoundingBox::FromLtrb(box_data)
-                     : BoundingBox::FromXywh(box_data);
-    bounding_boxes.emplace_back(box);
+      return ProspectiveCrop(true, out_crop, filtered_bboxes, filtered_labels);
+    }
+
+    return ProspectiveCrop();
+  }
+
+  bool ValidAspectRatio(span<const float> shape) {
+    assert(static_cast<int>(shape.size()) == coords_size);
+    for (int i = 0; i < ndim; i++) {
+      for (int j = i + 1; j < ndim; j++) {
+        if (!aspect_ratio_range_.Contains(shape[i] / shape[j]))
+          return false;
+      }
+    }
+    return true;
+  }
+
+  bool ValidOverlap(const Box<ndim, float> &crop, const std::vector<Box<ndim, float>> &boxes,
+                    float threshold) {
+    return std::all_of(boxes.begin(), boxes.end(),
+      [&crop, threshold](const Box<ndim, float> &box) {
+        return intersection_over_union(crop, box) >= threshold;
+      });
+  }
+
+  void FilterByCentroid(const Box<ndim, float> &crop,
+                        std::vector<Box<ndim, float>> &bboxes,
+                        std::vector<int> &labels) {
+    std::vector<Box<ndim, float>> new_bboxes;
+    std::vector<int> new_labels;
+    bool process_labels = !labels.empty();
+    assert(labels.empty() || labels.size() == bboxes.size());
+    for (size_t i = 0; i < bboxes.size(); i++) {
+      if (crop.contains(bboxes[i].centroid())) {
```

Comment:
Beware that `hi` edge is not considered to lie inside `Box` - do we want that? Do we care?